### PR TITLE
Feature #169954 chore: Add training records in bulk

### PR DESCRIPTION
**Pull request by tushar33**

### CODE REVIEW

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+			$userdata['name']         = $data['name'];
+			$userdata['issuing_org']  = $data['issuing_org'];
+			$userdata['issued_on']    = $data['issued_on'];
+			$userdata['expired_on']   = $data['expired_on'];
+			$userdata['status']       = $data['status'];
+			$userdata['created_by']   = $data['created_by'];
+			$userdata['notify']       = $data['notify_users'];
+			$userdata['state']        = $data['state'];
+			$userdata['client']       = "external";
+			$userdata['is_external']  = 1;
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: optimise code

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+					{
+						$agencyModel = TJCERT::model('Agency', array('ignore_request' => true));
+
+						// Get agencies of logged-in user and assigned user
+						$assignedUserAgencies = $agencyModel->getUserAgencies($userId);
+						$loggedInUserAgencies = $agencyModel->getUserAgencies($user->id);
+
+						// Convert object to array
+						$assignedUserAgencyArr = array_column($assignedUserAgencies, 'id');
+						$loggedInUserAgencyArr = array_column($loggedInUserAgencies, 'id');
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: Check if  object or array is empt

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+				$userdata['user_id'] = $userId;
+
+				if ($app->input->get('site', 'f', 'string') == 'f')
+				{
+					$redirectLink = Route::_('index.php?option=com_tjcertificate&view=certificates&layout=my', false);
+				}
+				else
+				{
+					$redirectLink = Route::_('index.php?option=com_tjcertificate&view=certificates', false);
+				}
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: We can remove this

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+					$returnData['msg']          = $msg;
+					$returnData['redirect_url'] = $redirectLink;
+				}
+				else
+				{
+					$certificateModel = TJCERT::model('Certificate', array('ignore_request' => true));
+					$certificateModel->save($userdata);
+					$certificateIds[] = $certificateModel->getState('certificate.id');
+
+					$returnData = array();
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: declare only once

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+					if ($certificateIds)
+					{
+						$error = false;
+						$returnData['msg']          = Text::sprintf('COM_TJCERTIFICATE_TOTAL_RECORDS_ADDED', count($certificateIds));
+						$returnData['redirect_url'] = $redirectLink;
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: you can remove

File: src/components/com_tjcertificate/administrator/libraries/certificate.php [hunk 6]
```diff
@@ -494,7 +496,7 @@ public function save()
 
 			$dispatcher = \JEventDispatcher::getInstance();
 
-			if ($table->is_external && $isNew)
+			if ($table->is_external && $isNew )
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/libraries/certificate.php]: remove extra space

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php (hunk 1 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: done

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php (hunk 3 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: removed

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php (hunk 4 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: done

File: src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php (hunk 5 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/trainingrecords.json.php]: removed

File: src/components/com_tjcertificate/administrator/libraries/certificate.php (hunk 6 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/libraries/certificate.php]: removed

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php [hunk 7]
```diff
@@ -0,0 +1,152 @@
+<?php
+/**
+ * @package     TJCertificate
+ * @subpackage  com_tjcertificate
+ *
+ * @author      Techjoomla <extensions@techjoomla.com>
+ * @copyright   Copyright (C) 2009 - 2020 Techjoomla. All rights reserved.
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: 2021

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	protected $comMultiAgency = 'com_multiagency';
+
+	/**
+	 * Function to add multiple records
+	 *
+	 * @return  void
+	 *
+	 * @since  __DEPLOY_VERSION__
+	 */
+	public function addRecords()
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: rename it to save()

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+					if ($manageOwn && empty($manage))
+					{
+						$agencyModel = TJCERT::model('Agency', array('ignore_request' => true));
+
+						// Get agencies of logged-in user and assigned user
+						$assignedUserAgencies = $agencyModel->getUserAgencies($userId);
+						$loggedInUserAgencies = $agencyModel->getUserAgencies($user->id);
+
+						$result = '';
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: $result = array();

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+					}
+				}
+
+				$data['user_id'] = $userId;
+				$returnData = array();
+
+				if (ComponentHelper::isEnabled('com_tjqueue') && $params->get('tjqueue_records'))
+				{
+					$recordsModel      = TJCERT::model('BulkTrainingRecord', array('ignore_request' => true));
+					$response          = $recordsModel->queueRecords($data);
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: change function name to addToQueue($data)

File: src/components/com_tjcertificate/administrator/controllers/trainingrecord.php [hunk 11]
```diff
@@ -111,7 +111,13 @@ public function save($key = null, $urlVar = null)
 		}
 
 		$validData['client'] = "external";
-		$validData['state'] = $validData['state'] ? $validData['state'] : "-1";
+
+		// Don't update state on frontend edit
+		if (!$validData['id'])
+		{
+			$validData['state'] = $validData['state'] ? $validData['state'] : "-1";
+		}
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/trainingrecord.php]: Can you explain this code?

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini [hunk 12]
```diff
@@ -383,3 +383,24 @@ COM_TJCERTIFICATE_PERMISSION_DELETE_EXTERNAL_CERTIFICATE="Delete training record
...
+COM_TJCERTIFICATE_TJQUEUE_SETTING="Use TjQueue while add bulk records"
+COM_TJCERTIFICATE_TJQUEUE_SETTING_DESC="Set to 'Yes' if you want to use TjQueue while add bulk records (Note: To work this functionality you need com_tjqueue and certificate plugin enable)"
+COM_TJCERTIFICATE_RECORDS_USER_SELECT_LIMIT="User limit for add bulk records"
+COM_TJCERTIFICATE_RECORDS_USER_SELECT_LIMIT_DESC="This limit used only when you are not using TjQueue and limit above 40 will result timeout."
+COM_TJCERTIFICATE_USER_LIMIT_MESSAGE="You can not add records more than %s user"
+COM_TJCERTIFICATE_BULK_RECORD_SETTINGS="Bulk Training Record"
+COM_TJCERTIFICATE_BULK_RECORD_SETTINGS_NOTE="Please note: <br> <br> <label>You need to follow below steps to make it work properly with <i>TjQueue</i>: <br><br> 1. TjQueue extension must be installed in your site <br> 2. Enable <i>TjQueue plugin for TjCertificate</i> <br> 3. Find a information below to setup a cron <br> <a href="https://github.com/techjoomla/com_tjqueue/wiki/Tj-Queue-Integration#cron-setup" target="_blank">Cron Setup
+</a><br><br> Example: <br> */1 * * * * /usr/bin/php7.3 JOOMLA_SITE/cli/tjqueue.php -t tjqueue -n 5
+</label>"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini:396]: There is a syntax error here. Use a single quote for the anchor tag link

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini [hunk 13]
```diff
@@ -383,3 +383,24 @@ COM_TJCERTIFICATE_PERMISSION_DELETE_EXTERNAL_CERTIFICATE="Delete training record
 COM_TJCERTIFICATE_PERMISSION_DELETE_EXTERNAL_CERTIFICATE_DESC="Allow users to delete training records"
 COM_TJCERTIFICATE_PERMISSION_DELETEOWN_EXTERNAL_CERTIFICATE="Delete own records"
 COM_TJCERTIFICATE_PERMISSION_DELETEOWN_EXTERNAL_CERTIFICATE_DESC="Allow users to delete own training records"
+
+;Bulk Record Config
+COM_TJCERTIFICATE_TJQUEUE_SETTING="Use TjQueue while add bulk records"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: Use TjQueue while adding bulk records

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini [hunk 14]
```diff
@@ -383,3 +383,24 @@ COM_TJCERTIFICATE_PERMISSION_DELETE_EXTERNAL_CERTIFICATE="Delete training record
...
+COM_TJCERTIFICATE_TJQUEUE_SETTING_DESC="Set to 'Yes' if you want to use TjQueue while add bulk records (Note: To work this functionality you need com_tjqueue and certificate plugin enable)"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: Set to 'Yes' if you want to use TjQueue while adding bulk records (Note: To make it work, you need com_tjqueue component, and also you need to enable certificate queue plugin)

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini [hunk 15]
```diff
@@ -383,3 +383,24 @@ COM_TJCERTIFICATE_PERMISSION_DELETE_EXTERNAL_CERTIFICATE="Delete training record
...
+COM_TJCERTIFICATE_RECORDS_USER_SELECT_LIMIT="User limit for add bulk records"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: User limit for adding bulk records

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini [hunk 16]
```diff
@@ -383,3 +383,24 @@ COM_TJCERTIFICATE_PERMISSION_DELETE_EXTERNAL_CERTIFICATE="Delete training record
...
+COM_TJCERTIFICATE_RECORDS_USER_SELECT_LIMIT_DESC="This limit used only when you are not using TjQueue and limit above 40 will result timeout."
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: This limit value is used only when you are not using TjQueue and if the limit is set above 40 may result in a timeout.

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini [hunk 17]
```diff
@@ -383,3 +383,24 @@ COM_TJCERTIFICATE_PERMISSION_DELETE_EXTERNAL_CERTIFICATE="Delete training record
...
+COM_TJCERTIFICATE_USER_LIMIT_MESSAGE="You can not add records more than %s user"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: You can not add more than the %s records.

File: src/components/com_tjcertificate/administrator/models/bulktrainingrecord.php [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+	 *
+	 * @param   array  $data  record data
+	 *
+	 * @return  boolean value.
+	 *
+	 * @since __DEPLOY_VERSION__
+	 */
+	public function queueRecords($data)
+	{
+		$return      = [];
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/bulktrainingrecord.php]: $return      = array();

File: src/components/com_tjcertificate/administrator/views/bulktrainingrecord/tmpl/edit.php [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+		</div>
+		<input type="hidden" name="jform[created_by]" value="<?php echo Factory::getUser()->id;?>" />
+		<input type="hidden" name="task" value="bulktrainingrecord.addRecords" />
+		<?php echo HTMLHelper::_('form.token'); ?>
+	</div>
+	</form>
+</div>
+</div>
+<script>
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/views/bulktrainingrecord/tmpl/edit.php]: <script type="text/javascript">

File: src/components/com_tjcertificate/administrator/views/bulktrainingrecord/tmpl/edit.php [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+	var userLimit = '<?php echo $userLimit;?>';
+	var message = '<?php echo $message;?>';
+
+	jQuery("#jform_assigned_user_id").chosen({max_selected_options: userLimit});
+	jQuery("#jform_assigned_user_id").bind("liszt:maxselected", function () {
+		Joomla.renderMessages({'error': [message] });
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/views/bulktrainingrecord/tmpl/edit.php:91]: please show in a demo, how it looks

File: src/components/com_tjcertificate/administrator/views/bulktrainingrecord/view.html.php [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected function addToolbar()
+	{
+		$app = Factory::getApplication();
+
+		$layout = $app->input->get("layout");
+
+		JLoader::import('administrator.components.com_tjcertificate.helpers.tjcertificate', JPATH_SITE);
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/views/bulktrainingrecord/view.html.php:79]: You don't need to import this

File: src/components/com_tjcertificate/site/views/bulktrainingrecord/tmpl/edit.php [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+					<span><?php echo Text::_('JCANCEL'); ?></span>
+			</button>
+		</div>
+	</div>
+	<input type="hidden" name="option" value="com_tjcertificate"/>
+	<input type="hidden" name="task" value="bulktrainingrecord.addRecords"/>
+	<input type="hidden" name="site" value="f"/>
+	<?php echo HTMLHelper::_('form.token'); ?>
+</form>
+<script>
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/site/views/bulktrainingrecord/tmpl/edit.php]: <script type="text/javascript"></script>

Comment (tushar33) [reply src/components/com_tjcertificate/site/views/bulktrainingrecord/tmpl/edit.php]: added

File: src/components/com_tjcertificate/administrator/views/bulktrainingrecord/view.html.php (hunk 21 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/views/bulktrainingrecord/view.html.php:79]: This is needed to show the side menu

File: src/components/com_tjcertificate/administrator/views/bulktrainingrecord/tmpl/edit.php (hunk 19 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/views/bulktrainingrecord/tmpl/edit.php]: added

File: src/components/com_tjcertificate/administrator/models/bulktrainingrecord.php (hunk 18 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/bulktrainingrecord.php]: added

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini (hunk 17 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: updated

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini (hunk 16 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: We are using limit with or without tjqueue so updating text as below

If the limit is set above 40 may result in a timeout.

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini (hunk 15 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: updated

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini (hunk 13 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: updated

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini (hunk 14 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini]: updated

File: src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini (hunk 12 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/languages/en-GB/en-GB.com_tjcertificate.ini:396]: Updated

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php (hunk 7 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: updated

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php (hunk 9 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: updated

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php (hunk 10 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: updated

File: src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php (hunk 8 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/bulktrainingrecord.json.php]: updated

File: src/components/com_tjcertificate/plugins/tjqueue/certificate/certificate.php [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+use Joomla\CMS\Plugin\CMSPlugin;
+
+/**
+ * TjQueue
+ *
+ * @package     Techjoomla.Libraries
+ * @subpackage  Tjqueue
+ * @since       1.0
+ */
+class PlgTjqueueRecords extends CMSPlugin
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/plugins/tjqueue/certificate/certificate.php:21]: Plugin name and plugin class name is different?

Comment (tushar33) [reply src/components/com_tjcertificate/plugins/tjqueue/certificate/certificate.php:21]: this type of structure needed in cli, check following ref code of cli

$filePath = JPATH_SITE . '/plugins/tjqueue/' . $plugin . '/consumers/' . $class . '.php';

**Pull request merged**